lib/transport/request.js

```diff
@@ -1,72 +1,107 @@
 'use strict';
 
-const _ = require('lodash');
-const promisifyAll = require('./promiseUtils').promisifyAll;
 const Transport = require('./transport');
-const Request = require('request');
-const dns = require('dns');
+const https = require('node:https');
+// eslint-disable-next-line func-style
+const fetch = (...args) => import('node-fetch').then(({ default: fetch }) => fetch(...args));
 
 const REQUIRED_PROPERTIES = [
   'body',
   'elapsedTime',
   'url',
   'statusCode',
+  'statusText',
   'headers'
 ];
 
-function toAsyncMethod(method) {
-  return `${method.toLowerCase()}Async`;
-}
-
 class RequestTransport extends Transport {
-  constructor(customRequest) {
+  constructor(defaultOptions) {
     super();
-    this._request = promisifyAll(customRequest || Request);
+    if (defaultOptions?.agentOpts) {
+      this._agent = new https.Agent(defaultOptions.agentOpts);
+    }
+
+    this._fetch = fetch;
   }
 
   toOptions(ctx) {
     const req = ctx.req;
-    const opts = Object.assign({
-      time: true,
-      agentOptions: {
-        lookup: function(domain, options, cb) {
-          // Prevent Node from reordering A and AAAA records.
-          // See https://nodejs.org/api/dns.html#dns_dns_lookup_hostname_options_callback
-          options.verbatim = true;
-          return dns.lookup(domain, options, cb);
-        }
-      }
-    }, ctx.opts);
+    const opts = ctx.opts || {};
 
     if (req.getTimeout() !== undefined) {
       opts.timeout = req.getTimeout();
     }
 
-    if (!_.isUndefined(req.time)) opts.time = req.time;
-    if (req.hasQueries()) opts.qs = req.getQueries();
-    if (req.hasHeaders()) opts.headers = req.getHeaders();
+    if (req.hasQueries()) opts.searchParams = new URLSearchParams(req.getQueries());
+
     const body = ctx.req.getBody();
     if (body) {
-      if (typeof body === 'object') opts.json = true;
-      opts.body = body;
+      if (typeof body === 'object') {
+        req.addHeader('Content-type', 'application/json');
```

Comment:
Actually I just found some middleware that handles json, is that the better pattern? Rather than introducing complexity here?

Reply:
This code has the same behaviour as before, by setting `opts.json = true;` you are telling `request` that the content type of the request body AND response body is json.

Reply:
`request` would also parse the response body as json if `opts.json = true;` so the middleware is actually redundant if you set this option

Reply:
We are now going off the response headers to parse text or json [here](https://github.com/bbc/http-transport/blob/1ff8ac9cc62f64e50dd5f5cd54f902d63f840bd9/lib/transport/request.js#L63C43-L63C43)

Reply:
however, it could be required that you use the correct middleware to process your response. This kind of defeats the point of it being middleware however

Reply:
> This code has the same behaviour as before, by setting `opts.json = true;` you are telling `request` that the content type of the request body AND response body is json.

I'm going to take it that the last `AND response body is json` isn't right then.